// steam-binary-vdf/out/BinaryVdf.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getShortcutUrl = exports.getShortcutHash = exports.writeVdf = exports.readVdf = void 0;
const buffer_reader_1 = __importDefault(require("buffer-reader"));
const polycrc = __importStar(require("polycrc"));
const node_int64_1 = __importDefault(require("node-int64"));
const hex2dec_1 = require("hex2dec");
function nextMapItem(buffer) {
    const typeByte = buffer.nextUInt8();
    if (typeByte === 8 /* MapEnd */) {
        return {
            type: typeByte,
        };
    }
    const name = buffer.nextStringZero();
    let value;
    switch (typeByte) {
        case 0 /* Map */: {
            value = nextMap(buffer);
            break;
        }
        case 1 /* String */: {
            value = buffer.nextStringZero("utf-8");
            break;
        }
        case 2 /* Number */: {
            value = buffer.nextUInt32LE();
            break;
        }
    }
    return {
        type: typeByte,
        name: name,
        value: value,
    };
}
function nextMap(buffer) {
    const contents = {};
    while (true) {
        const mapItem = nextMapItem(buffer);
        if (mapItem.type === 8 /* MapEnd */) {
            break;
        }
        contents[mapItem.name] = mapItem.value;
    }
    return contents;
}
function readVdf(buffer, offset) {
    const reader = new buffer_reader_1.default(buffer);
    if (offset) {
        reader.seek(offset);
    }
    return nextMap(reader);
}
exports.readVdf = readVdf;
function addByte(value, contents) {
    const typeBuffer = Buffer.allocUnsafe(1);
    typeBuffer.writeUInt8(value);
    contents.push(typeBuffer);
}
function addNumber(value, contents) {
    const valueBuffer = Buffer.allocUnsafe(4);
    valueBuffer.writeUInt32LE(value);
    contents.push(valueBuffer);
}
function addString(value, contents) {
    if (value.indexOf("\0") !== -1) {
        throw new Error('Strings in VDF files cannot have null chars ("\\0")');
    }
    contents.push(Buffer.from(value, "utf-8"));
    addByte(0, contents);
}
function addMap(map, contents) {
    for (const key of Object.keys(map)) {
        const value = map[key];
        if (typeof value === "number") {
            addByte(2 /* Number */, contents);
            addString(key, contents);
            addNumber(value, contents);
        }
        else if (typeof value === "string") {
            addByte(1 /* String */, contents);
            addString(key, contents);
            addString(value, contents);
        }
        else if (typeof value === "object") {
            addByte(0 /* Map */, contents);
            addString(key, contents);
            addMap(value, contents);
        }
        else {
            throw new Error(`Type at ${key} (${typeof value}) is not allowed in VDF files. VDF files can only contain numbers, strings, or objects`);
        }
    }
    const endBuffer = Buffer.allocUnsafe(1);
    endBuffer.writeUInt8(8 /* MapEnd */);
    contents.push(endBuffer);
}
function writeVdf(map) {
    const contents = [];
    addMap(map, contents);
    return Buffer.concat(contents);
}
exports.writeVdf = writeVdf;
let shortcutCrc;
function getShortcutHash(input) {
    shortcutCrc = shortcutCrc || polycrc.crc(32, 0x04c11db7, 0xffffffff, 0xffffffff, true);
    const full64 = new node_int64_1.default(shortcutCrc(input) | 0x80000000, 0x02000000);
    return hex2dec_1.hexToDec(full64.toOctetString());
}
exports.getShortcutHash = getShortcutHash;
function getShortcutUrl(appName, exe) {
    return "steam://rungameid/" + getShortcutHash(exe + appName);
}
exports.getShortcutUrl = getShortcutUrl;
/*
getShortcutHash uses code adapted from [Scott Rice's ICE
program](https://github.com/scottrice/Ice) which is under the following license:

Copyright (c) 2012, 2013 Scott Rice
All rights reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
